src/GNDStk/gnds-convert-to-json.hpp

```diff
@@ -0,0 +1,134 @@
+
+/*
+Summary of the functions in this file:
+
+namespace detail {
+   1. prefix(unsigned long)
+
+   2. Node2json(gnds::Node, nlohmann::json, unsigned long)
+      ...uses (1)
+}
+
+3. convert(gnds::Tree, gnds::json)
+   ...uses (2)
+
+4. convert(gnds::xml,  gnds::json)
+   ...uses (3)
+
+5. convert(gnds::json, gnds::json)
+*/
+
+
+
+// -----------------------------------------------------------------------------
+// Helpers
+// -----------------------------------------------------------------------------
+
+namespace detail {
+
+// prefix
+inline std::string prefix(const unsigned long n)
+{
+   std::ostringstream oss;
+   oss << std::setfill('0') << std::setw(12) << n;
+   return oss.str();
+}
+
+
+
+// Node2json
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool Node2json(
+   const gnds::Node<MCON,CCON> &node,
+   nlohmann::json &j,
+   unsigned long &kwdcount
+) {
+   // name
+   // The effect of j[nodename] is to enter a key of this name into the json
+   // object. Generally, this is triggered automatically in the body of one
+   // and/or the other of the upcoming metadata and children loops. However,
+   // consider a node that has no metadata or children; then, we need this.
+   // An example is <RutherfordScattering/> in some of our GNDS XML files,
+   // as it contains neither metadata nor children.
+   const std::string nodename = prefix(kwdcount++) + node.name;
+   j[nodename];
+
+   // metadata
+   if (node.metadata.size() > 0) {
+      // ...because we only want the kwdcount ++ side effect if size() > 0
+      const std::string attrname = prefix(kwdcount++) + "attributes";
+      // visit
+      for (auto &meta : node.metadata)
+         j[nodename][attrname][prefix(kwdcount++) + meta.first] =
+            meta.second;
+   }
+
+   // children
+   for (auto &child : node.children)
+      if (child && !Node2json(*child, j[nodename], kwdcount))
+         return false;
+
+   // done
+   return true;
+}
+
+} // namespace detail
+
+
+
+// -----------------------------------------------------------------------------
+// convert
+// -----------------------------------------------------------------------------
+
+// Tree ==> json
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(const gnds::Tree<MCON,CCON> &tree, gnds::json &jdoc)
```

Comment:
I can certainly go ahead and provide both an overloaded convert() for these (well, that's what I already have), plus functions of the form foo2bar for each appropriate foo and bar. I'd suggest that the "foo2bar" notation is actually an older and somewhat deprecated methodology for doing this sort of thing. It was popular in C, back in the pre-C++ days, because function overloading didn't exist at the time; every function needed a different name. Now, in the same way that the standard C++ library has a sqrt() function with overloads for various argument types - as opposed to, say, fsqrt() for floats, dsqrt() for double, etc. - I'm doing the same with convert(). I can certainly put in the foo2bar forms of the functions, but my guess is that most users will simply use convert().